LoopKit/TemporaryScheduleOverrideHistory.swift

```diff
@@ -38,66 +43,102 @@ public protocol TemporaryScheduleOverrideHistoryDelegate: AnyObject {
+    }
 
     public func recordOverride(_ override: TemporaryScheduleOverride?, at enableDate: Date = Date()) {
-        guard override != recentEvents.last?.override else {
+        guard override != lastUndeletedEvent?.override else {
             return
         }
-
+        
         if let override = override {
             record(override, at: enableDate)
         } else {
             cancelActiveOverride(at: enableDate)
         }
     }
-
+    
+    private var lastUndeletedEvent: OverrideEvent? {
+        return recentEvents.reversed().first { $0.end != .deleted }
+    }
+    
+    private func deleteEventsStartingOnOrAfter(_ date: Date) {
+        recentEvents.mutateEach { (event) in
+            if event.override.startDate >= date {
+                event.end = .deleted
+                event.modificationCounter = modificationCounter
+            }
+        }
+    }
+    
+    // Deletes overrides that start after the passed in override.
     private func record(_ override: TemporaryScheduleOverride, at enableDate: Date) {
-        recentEvents.removeAll(where: { $0.override.startDate >= override.startDate })
-
-        if  let lastEvent = recentEvents.last,
-            case let overrideEnd = min(override.startDate.nearestPrevious, enableDate),
-            lastEvent.actualEndDate > overrideEnd
-        {
-            recentEvents[recentEvents.endIndex - 1].end = .early(overrideEnd)
+        
+        // Check for modification of existing entry
+        var index = recentEvents.endIndex
+        while index != recentEvents.startIndex {
+            recentEvents.formIndex(before: &index)
+            if recentEvents[index].override.syncIdentifier == override.syncIdentifier {
+                recentEvents[index].override = override
+                recentEvents[index].modificationCounter = modificationCounter
+                return
+            }
         }
-
-        let enabledEvent = OverrideEvent(override: override)
+        
+        deleteEventsStartingOnOrAfter(override.startDate)
+        
+        let overrideEnd = min(override.startDate.nearestPrevious, enableDate)
+        
+        cancelActiveOverride(at: overrideEnd)
+
+        let enabledEvent = OverrideEvent(override: override, modificationCounter: modificationCounter)
         recentEvents.append(enabledEvent)
     }
 
     private func cancelActiveOverride(at date: Date) {
-        if  let lastEvent = recentEvents.last,
-            case .natural = lastEvent.end,
-            !lastEvent.override.hasFinished(relativeTo: date)
-        {
-            if lastEvent.override.startDate > date {
-                recentEvents.removeLast()
-            } else {
-                recentEvents[recentEvents.endIndex - 1].end = .early(date)
+        var index = recentEvents.endIndex
+        while index != recentEvents.startIndex {
+            recentEvents.formIndex(before: &index)
+            
+            if recentEvents[index].end != .deleted {
```

Comment:
This searches for non-`.deleted` events, but should this include `.early` or `.natural`? Or are such events still considered active? (I'm probably missing something)

Reply:
Yes, .early and .natural end types are valid for cancelation. Most events being canceled will have a .natural end. Events with an .early end could still be canceled if the cancel at date is before the actualEndDate.